### DISCUSSION TITLE
Supporting other languages by providing interpreter for JSON structure

### DISCUSSION THREAD

**Discussion by marcustut** [statelyai/xstate#3836] [General]
XState is fantastic, ever since I've been introduced to it while doing frontend react, I've been trying to use it in a lot of my other application. I absolutely love the fact that I can model my statecharts using stately or the VS Code extension. 

Recently, I've been working in Rust and started wondering since statecharts in XState are just represented in JSON format, if we could build an interpreter or a library that can understand and parse the JSON format in other languages, doesn't that mean we can use XState in other languages as well?

**Comment by davidkpiano** [comment]
Yes, it does! The interpreter is the hard part, of course, but that is what the JSON structure enables.

**Reply by marcustut** [reply]
I'm not sure if this idea is valid but could be a big game changer if it can be done. 

WASM is on the rise and a lot of the modern languages has support for it. Rust can load WASM modules through runtimes like Wasmer or WasmTime and there's emscripten for C/C++.

Considering the similarities between AssemblyScript and TypeScript, I would imagine that it is the least effort factoring in the options, I did saw several other discussions regarding AssemblyScript here, hope it will be a thing in the future as the project gets more attention and have more contributors.

**Reply by davidkpiano** [reply]
It's definitely an interesting idea. There's considerable effort in porting XState v5 to a different language, but it is possible.
